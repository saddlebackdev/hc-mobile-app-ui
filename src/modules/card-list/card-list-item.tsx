// Modules
import * as React from 'react';
import Styled from 'styled-components/native';

// Types
import {IProps} from './card-list-item.types';

// Shared
import Text from '../text/text';
import Heading from '../heading/heading';
import {majorScale} from '../scales';

// Styles
const StyledCard = Styled.TouchableOpacity`
  flex-direction: column;
  padding-vertical: ${majorScale(2)}px;
`;
const StyledCardContent = Styled.View`
  flex-direction: row;
`;
const StyledCardMarker = Styled.View`
  margin-top: ${majorScale(1)}px;
`;
const StyledCardPhotoWrapper = Styled.View`
  width: 35%;
`;
const StyledCardDetailsWrapper = Styled.View`
  width: 65%;
`;
const StyledCardDetailsRow = Styled.View``;
const StyledPhoto = Styled.Image`
  width: 96px; height: 96px;
  border-radius: 8px;
`;
const StyledTitle = Styled(Heading)``;
const StyledSubTitle = Styled(Text)`
  font-size: 12px;
  margin-top: 8px;
`;
const StyledDate = Styled(Text)`
  margin-top: 8px;
`;
const Styledtime = Styled(Text)`
  margin-top: 8px;
`;
const StyledDescription = Styled(Text)`
  margin-top: 8px;
  font-size: ${({theme}) => theme.typography.sizes.small}px;
`;
const StyledTags = Styled(Text)`
  margin-top: 8px;
`;
// Component
export const CardListItem: React.FC<IProps> = ({
  photoUrl,
  fallbackImage,
  title,
  marker = null,
  timePeriod = null,
  subTitle,
  eventDate,
  eventTime,
  description,
  tags,
  onPress,
}): React.ReactElement => (
  <React.Fragment>
    <StyledCard activeOpacity={0.75} onPress={onPress} testID="list-item">
      <StyledCardContent>
        <StyledCardPhotoWrapper>
          <StyledPhoto
            source={photoUrl ? {uri: photoUrl} : fallbackImage}
            testID="item-photo"
          />
        </StyledCardPhotoWrapper>
        <StyledCardDetailsWrapper>
          <StyledCardDetailsRow>
            <StyledTitle variant="h4" testID="item-title">
              {title}
            </StyledTitle>
            {subTitle ? (
              <StyledSubTitle variant="subtitle2" testID="item-subtitle">
                {subTitle}
              </StyledSubTitle>
            ) : null}
            {eventDate ? (
              <StyledDate
                weight="bold"
                muted
                variant="subtitle2"
                testID="item-eventdate">
                {eventDate}
                {eventTime ? (
                  <Styledtime variant="subtitle2" muted testID="item-eventtime">
                    {eventTime}
                  </Styledtime>
                ) : null}
              </StyledDate>
            ) : null}
            {timePeriod ? (
              <StyledCardMarker>{timePeriod}</StyledCardMarker>
            ) : null}
          </StyledCardDetailsRow>
          {description ? (
            <StyledCardDetailsRow>
              <StyledDescription
                numberOfLines={2}
                variant="subtitle2"
                muted
                testID="item-description">
                {description}
              </StyledDescription>
            </StyledCardDetailsRow>
          ) : null}
          {tags ? (
            <StyledCardDetailsRow>
              <StyledTags
                muted
                italic
                variant="subtitle2"
                weight="bold"
                testID="item-tags">
                {tags.map((tag, ndx) => {
                  const isLastChild = tags.length - 1 === ndx;

                  return isLastChild ? tag : `${tag} `;
                })}
              </StyledTags>
            </StyledCardDetailsRow>
          ) : null}
        </StyledCardDetailsWrapper>
      </StyledCardContent>

      {marker ? <StyledCardMarker>{marker}</StyledCardMarker> : null}
    </StyledCard>
  </React.Fragment>
);
// Exports
export default CardListItem;
