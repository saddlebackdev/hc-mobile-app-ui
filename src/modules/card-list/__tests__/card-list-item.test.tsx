// npx jest ./src/modules/lists/__tests__/card-list-item.test.ts

import * as React from 'react';
import {fireEvent, render} from '@testing-library/react-native';

import {CardListItem} from '../card-list-item';
import ThemeProvider from '../../theming/theme-provider';
import defaultTheme from '../../theming/default-theme';

const onPressMock = jest.fn();
const cardListItems = [
  {
    id: 'card-list-item-1',
    title: 'A Title',
    subTitle: 'A Subtitle',
    eventDate: 'A eventDate',
    eventTime: 'A eventTime',
    photoUrl: 'https://images.com/abc.png',
    description: 'A Description',
    tags: ['Tag One', 'Tag Two'],
    onPress: onPressMock,
  },
  {
    id: 'card-list-item-2',
    title: 'A Title',
    subTitle: 'A Subtitle',
    eventDate: 'A eventDate',
    eventTime: 'A eventTime',
    photoUrl: 'https://images.com/abc.png',
    description: 'A Description',
    tags: null,
    onPress: onPressMock,
  },
];

describe('CardListItem', () => {
  const [item] = cardListItems;

  describe('Rendering', () => {
    it('renders itself correctly', () => {
      const wrapper = render(
        <ThemeProvider theme={defaultTheme}>
          <CardListItem {...item} />
        </ThemeProvider>,
      );

      const listItem = wrapper.getByTestId('list-item');

      expect(listItem).toBeDefined();

      expect(wrapper).toMatchSnapshot();
    });

    it('hides tags when it is null', () => {
      const wrapper = render(
        <ThemeProvider theme={defaultTheme}>
          <CardListItem {...item} tags={null} />
        </ThemeProvider>,
      );

      const itemTags = wrapper.queryByTestId('item-tags');

      expect(itemTags).toBeNull();

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Interactions', () => {
    it('should call onPress method when pressed', () => {
      const wrapper = render(
        <ThemeProvider theme={defaultTheme}>
          <CardListItem {...item} />
        </ThemeProvider>,
      );

      const listItem = wrapper.getByTestId('list-item');

      expect(listItem).toBeDefined();

      fireEvent.press(listItem);

      expect(onPressMock).toHaveBeenCalledTimes(1);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
