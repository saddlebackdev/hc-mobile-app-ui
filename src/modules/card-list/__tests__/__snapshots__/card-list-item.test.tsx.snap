// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardListItem Interactions should call onPress method when pressed 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flexDirection": "column",
      "opacity": 1,
      "paddingVertical": 22,
    }
  }
  testID="list-item"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": "35%",
          },
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "https://images.com/abc.png",
          }
        }
        style={
          Array [
            Object {
              "borderRadius": 8,
              "height": 96,
              "width": 96,
            },
          ]
        }
        testID="item-photo"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "width": "65%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1C2530",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "700",
                "textAlign": "left",
              },
            ]
          }
          testID="item-title"
          variant="h4"
        >
          A Title
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#1C2530",
                "fontFamily": "System",
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-subtitle"
          variant="subtitle2"
        >
          A Subtitle
        </Text>
        <Text
          muted={true}
          style={
            Array [
              Object {
                "color": "#97A4AB",
                "fontFamily": "System",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "700",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-eventdate"
          variant="subtitle2"
          weight="bold"
        >
          A eventDate
          <Text
            muted={true}
            style={
              Array [
                Object {
                  "color": "#97A4AB",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "marginTop": 8,
                  "textAlign": "left",
                },
              ]
            }
            testID="item-eventtime"
            variant="subtitle2"
          >
            A eventTime
          </Text>
        </Text>
      </View>
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          muted={true}
          numberOfLines={2}
          style={
            Array [
              Object {
                "color": "#97A4AB",
                "fontFamily": "System",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-description"
          variant="subtitle2"
        >
          A Description
        </Text>
      </View>
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          italic={true}
          muted={true}
          style={
            Array [
              Object {
                "color": "#97A4AB",
                "fontFamily": "System",
                "fontSize": 14,
                "fontStyle": "italic",
                "fontWeight": "700",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-tags"
          variant="subtitle2"
          weight="bold"
        >
          Tag One 
          Tag Two
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`CardListItem Rendering hides tags when it is null 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flexDirection": "column",
      "opacity": 1,
      "paddingVertical": 22,
    }
  }
  testID="list-item"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": "35%",
          },
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "https://images.com/abc.png",
          }
        }
        style={
          Array [
            Object {
              "borderRadius": 8,
              "height": 96,
              "width": 96,
            },
          ]
        }
        testID="item-photo"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "width": "65%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1C2530",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "700",
                "textAlign": "left",
              },
            ]
          }
          testID="item-title"
          variant="h4"
        >
          A Title
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#1C2530",
                "fontFamily": "System",
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-subtitle"
          variant="subtitle2"
        >
          A Subtitle
        </Text>
        <Text
          muted={true}
          style={
            Array [
              Object {
                "color": "#97A4AB",
                "fontFamily": "System",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "700",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-eventdate"
          variant="subtitle2"
          weight="bold"
        >
          A eventDate
          <Text
            muted={true}
            style={
              Array [
                Object {
                  "color": "#97A4AB",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "marginTop": 8,
                  "textAlign": "left",
                },
              ]
            }
            testID="item-eventtime"
            variant="subtitle2"
          >
            A eventTime
          </Text>
        </Text>
      </View>
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          muted={true}
          numberOfLines={2}
          style={
            Array [
              Object {
                "color": "#97A4AB",
                "fontFamily": "System",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-description"
          variant="subtitle2"
        >
          A Description
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`CardListItem Rendering renders itself correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flexDirection": "column",
      "opacity": 1,
      "paddingVertical": 22,
    }
  }
  testID="list-item"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": "35%",
          },
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "https://images.com/abc.png",
          }
        }
        style={
          Array [
            Object {
              "borderRadius": 8,
              "height": 96,
              "width": 96,
            },
          ]
        }
        testID="item-photo"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "width": "65%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1C2530",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "700",
                "textAlign": "left",
              },
            ]
          }
          testID="item-title"
          variant="h4"
        >
          A Title
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#1C2530",
                "fontFamily": "System",
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-subtitle"
          variant="subtitle2"
        >
          A Subtitle
        </Text>
        <Text
          muted={true}
          style={
            Array [
              Object {
                "color": "#97A4AB",
                "fontFamily": "System",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "700",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-eventdate"
          variant="subtitle2"
          weight="bold"
        >
          A eventDate
          <Text
            muted={true}
            style={
              Array [
                Object {
                  "color": "#97A4AB",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "marginTop": 8,
                  "textAlign": "left",
                },
              ]
            }
            testID="item-eventtime"
            variant="subtitle2"
          >
            A eventTime
          </Text>
        </Text>
      </View>
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          muted={true}
          numberOfLines={2}
          style={
            Array [
              Object {
                "color": "#97A4AB",
                "fontFamily": "System",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-description"
          variant="subtitle2"
        >
          A Description
        </Text>
      </View>
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          italic={true}
          muted={true}
          style={
            Array [
              Object {
                "color": "#97A4AB",
                "fontFamily": "System",
                "fontSize": 14,
                "fontStyle": "italic",
                "fontWeight": "700",
                "marginTop": 8,
                "textAlign": "left",
              },
            ]
          }
          testID="item-tags"
          variant="subtitle2"
          weight="bold"
        >
          Tag One 
          Tag Two
        </Text>
      </View>
    </View>
  </View>
</View>
`;
